### PR TITLE
Symcache from Dwarf

### PR DESCRIPTION
**Pull request by mitsuhiko**
This work in progress pull request adds support for building a symcache from dwarf data.

### CODE REVIEW

**Comment by mitsuhiko** [comment]
note: this currently uses some code copied from addr2line. In case those survive we need to attribute this correctly.

File: symcache/src/cache.rs [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+            }
+        }
+        Ok(rv)
+    }
+
+    fn get_segment<T>(&self, seg: &Seg<T>) -> Result<&[T]> {
+        let offset = seg.offset as usize + mem::size_of::<CacheFileHeader>();
+        let size = mem::size_of::<T>() * seg.len as usize;
+        unsafe {
+            Ok(mem::transmute(self.byteview.get_data(offset, size)?))
```
Comment (jan-auer) [inline symcache/src/cache.rs]: Why `mem::transmute` here? Shouldn't `Result<&'a [T]>` do it?

File: symcache/src/cache.rs [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+            line: line,
+            symbol: self.get_symbol(fun.symbol_id)?,
+            filename: self.get_segment_as_string(&file_record.filename)?,
+            comp_dir: self.get_segment_as_string(&file_record.comp_dir)?,
+        })
+    }
+
+    /// Given an address this looks up the symbol at that point.
+    ///
+    /// Because of inling information this returns a vector of zero or
```
Comment (jan-auer) [inline symcache/src/cache.rs]: "inlining" typo

File: symcache/src/writer.rs [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+        SymCacheWriter {
+            writer: writer,
+        }
+    }
+
+    pub fn write_object(&mut self, obj: &Object) -> Result<()> {
+        macro_rules! section {
+            ($sect:ident) => {{
+                let sect = obj.get_dwarf_section(DwarfSection::$sect)
+                    .ok_or(ErrorKind::MissingSection(
```
Comment (jan-auer) [inline symcache/src/writer.rs]: We had a `MissingDebugInfo` error kind. Probably we should merge this.

File: common/src/errors.rs [hunk 4]
```diff
@@ -47,6 +46,11 @@ error_chain! {
             description("unknown cache file version")
             display("unknown cache file version '{}'", version)
         }
+        /// Raised if a section is missing in an object file.
```
Comment (jan-auer) [inline common/src/errors.rs]: Same as `MissingDebugInfo`?

File: symcache/src/writer.rs [hunk 5]
```diff
@@ -0,0 +1,319 @@
...
+    range: Option<gimli::Range>,
+    lines: Lines,
+    comp_dir: Option<gimli::EndianBuf<'input, Endianness>>,
+    programs: Vec<Program<'input>>,
+    language: Option<gimli::DwLang>,
+}
+
+impl<'input> Unit<'input> {
+    fn try_parse(
+        debug_abbrev: &gimli::DebugAbbrev<gimli::EndianBuf<Endianness>>,
```
Comment (jan-auer) [inline symcache/src/writer.rs]: use gimli::{...}

**Pull request merged**